## Import libraries

In [ ]:
import pandas as pd
import numpy as np
import sqlite3 as sql
import sqlalchemy

IMPORTANT: *PLEASE IMPORT MOVIE.DB BEFORE RUNNING ALL CELLS*
Li

## Data Retrival
In the database, there are three tables named: `IMDB`, `MOVIEDB`, `BOXOFFICE`.

We will proceed to use `SELECT * FROM (TABLE NAME)` to retrive the all data from each table. Then, we assign them each separate dataframe. First, let us write a function `sql_select` to retrieve data from the database.

In [ ]:
def sql_select(db,qry):
      conn=sql.connect(db)
      cur=conn.cursor()
      result=cur.execute(qry)
      return result.fetchall()


Now, we write 3 queries to retrieve all data from each table and assign them to corresponding dataframes

In [ ]:
#Assigning IMDB table to a DataFrame
qry1='SELECT * FROM IMDB'
df1= sql_select('movie.db',qry1)
IMDB_DF=pd.DataFrame(df1)
IMDB_DF.columns=['Title', 'Release', 'Rating']

#Assigning MOVIEDB to a DataFrame
qry2='SELECT * FROM MOVIEDB'
df2= sql_select('movie.db',qry2)
MOVIEDB_DF=pd.DataFrame(df2)
MOVIEDB_DF.columns=['Title', 'Revenue', 'Runtime']

#Assigning BOXOFFICE to a DataFrame
qry3='SELECT * FROM BOXOFFICE'
df3= sql_select('movie.db',qry3)
BOXOFFICE_DF=pd.DataFrame(df3)
BOXOFFICE_DF.columns=['Rank', 'Title', 'WorldWide_Lifetime_Gross', 'Domestic_Lifetime_Gross', 'Domestic(%)' , 'Foreign_Lifetime_Gross', 'Foreign(%)', 'Year']


Next, in order to satisfy Requirement #4, we write queries to select a subset of the data from each table in our database.

##Querying IMDB Table
First, we write queries to investigate subsets of data from the IMDB table

In [ ]:
IMDB_qry1='SELECT * FROM IMDB WHERE Rating = 8'           #Selecting movies with rating=8
IMDB_qry2='SELECT * FROM IMDB WHERE Rating IN (8.0, 9.0)' #Selecting movies with rating in range 8.0 to 9.0
IMDB_qry3='SELECT * FROM IMDB WHERE Rating >= 9.0'        #Selecting movies with rating>9
IMDB_qry4='SELECT * FROM IMDB ORDER BY Release DESC'      #Ordering movies by largest to smallest year
IMDB_qry5='SELECT * FROM IMDB ORDER BY Title ASC'         #Ordering movies alphabetically

sql_select('movie.db',IMDB_qry2)                          #Executing Queries. User can specify which query to call by changing the second parameter

[('The Dark Knight', 2008, 9.0),
 ('The Godfather Part II', 1974, 9.0),
 ('12 Angry Men', 1957, 9.0),
 ('Ford v Ferrari', 2019, 8.0),
 ('Rocky', 1976, 8.0),
 ('Platoon', 1986, 8.0),
 ('Spotlight', 2015, 8.0),
 ('Stand by Me', 1986, 8.0),
 ('The Terminator', 1984, 8.0),
 ('Logan', 2017, 8.0),
 ('Ratatouille', 2007, 8.0),
 ('Rush', 2013, 8.0),
 ('Pather Panchali', 1955, 8.0),
 ('Network', 1976, 8.0),
 ('Into the Wild', 2007, 8.0),
 ('The Wizard of Oz', 1939, 8.0),
 ('Before Sunset', 2004, 8.0),
 ('Groundhog Day', 1993, 8.0),
 ('The Exorcist', 1973, 8.0),
 ('The Best Years of Our Lives', 1946, 8.0),
 ('The Incredibles', 2004, 8.0),
 ('To Be or Not to Be', 1942, 8.0),
 ('The Battle of Algiers', 1966, 8.0),
 ("Hachi: A Dog's Tale", 2009, 8.0),
 ('Jai Bhim', 2021, 8.0),
 ('La haine', 1995, 8.0),
 ('Pirates of the Caribbean: The Curse of the Black Pearl', 2003, 8.0),
 ('The Grapes of Wrath', 1940, 8.0),
 ('My Father and My Son', 2005, 8.0),
 ('Amores Perros', 2000, 8.0),
 ('Rebecca', 1940, 8.

##Querying MOVIEDB Table
Next, we write queries to retrieve a subset of data from MOVIEDB

In [ ]:
MOVIE_qry1='SELECT * FROM  MOVIEDB WHERE Revenue > 0 ORDER BY Runtime DESC'                    #Selecting movies with non-zero revenue and ordered by descending runtime
MOVIE_qry2='SELECT * FROM  MOVIEDB WHERE Revenue > 0 AND RUNTIME >60 ORDER BY Runtime DESC'    #Selecting movies with non-zero revenue, runtime>60 and ordered from largest to smallest
MOVIE_qry3='SELECT * FROM  MOVIEDB WHERE Revenue > 0 ORDER BY Title ASC'                       #Selecting movies with non-zero revenue, with title ordered alphabetically

df4=sql_select('movie.db',MOVIE_qry2)  #Executing Query 2

MOVIE_DB_subset=pd.DataFrame(df4)      #Creating a subset df from MOVIE_DB, where movies are at least one hour in length and have non-zero revenue
MOVIE_DB_subset.columns=["Title", "Revenue", "Runtime"]
MOVIE_DB_subset




Title    Revenue  Runtime
0              Gone with the Wind  402352579      238
1     Once Upon a Time in America    5472914      229
2              Lawrence of Arabia   69995385      228
3                         Ben-Hur  164000000      222
4                   Seven Samurai     346300      207
...                           ...        ...      ...
2297                  Blood Feast    4000000       67
2298           She Done Him Wrong    2200000       66
2299                        Dumbo    1600000       64
2300                The Black Cat     236000       63
2301                   L'Âge d'Or       7940       63

[2302 rows x 3 columns]

##Querying BOXOFFICE Table
Finally, we write queries that retrieve subsets of data from BOXOFFICE with different properties.

In [ ]:
BOXOFFICE_qry1='SELECT * FROM  BOXOFFICE ORDER BY Rank ASC LIMIT 1'         #Selecting the first ranked movie
BOXOFFICE_qry2='SELECT Title, Max(Dgross) FROM  BOXOFFICE'                  #Selecting the movie with highest domestic lifetime gross
BOXOFFICE_qry3='SELECT Title, Min(Fgross) FROM  BOXOFFICE'                  #Selecting the movie with minimum foreign lifetime gross
BOXOFFICE_qry4='SELECT Title, Min(DPercent) FROM  BOXOFFICE'                #Selecting the movie with lowest domestic gross percentage
BOXOFFICE_qry5='SELECT Title, Max(FPercent) FROM  BOXOFFICE'                #Selecting the movie with highest foreign gross percentage
sql_select('movie.db',BOXOFFICE_qry5)                                       #Executing Queries. User can specify which query to call by changing the second parameter




[('The Battle at Lake Changjin', 100.0)]

## Data Visualization
After successfully obtaining dataframes for each table, we will use the columns to produce 5 charts. We will use the `IMDB_DF` dataframe to plot the first chart, `MOVIE_DB_subset` to plot the second and third charts, and `BOXOFFICE_DB` to plot the fourth and fifth charts.

###1. Movie Ratings Distribution
First, we will create a histogram that illustrates how movie ratings are distributed. This allows us to see which ratings are most commmon and which are the least common for the top 200 movies on IMDB's website.

In [ ]:
import plotly.express as px

fig = px.histogram(IMDB_DF, x="Rating", nbins=20, color='Rating', title="Fig1. Movie Ratings Distribution")
fig.update_layout(barmode='overlay')
# Reduce opacity to see both histograms
fig.update_traces(opacity=0.75)

fig.show()

###2. Runtime vs Revenue Scatterplot/Histogram
Next, we create a scatter plot with Runtime as the independent variable and Revenue as the dependent variable. Above the scatter plot, we also display its histogram representation

In [ ]:
fig = px.scatter(MOVIE_DB_subset, x="Runtime", y="Revenue", title="Fig2. Runtime vs Revenue", color='Runtime', marginal_x="histogram")

# Reduce opacity to see both histograms
fig.update_traces(opacity=0.75)

fig.show()

###3. Runtime vs Revenue Regression
Additionally, we also create a separate scatterplot of the same graph in 2., but this time with a regression line to model the relationship between runtime and revenue

In [ ]:
fig = px.scatter(MOVIE_DB_subset, x="Runtime", y="Revenue", title="Fig3. Runtime vs Revenue", color='Runtime', trendline="ols")

# Reduce opacity to see both histograms
fig.update_traces(opacity=0.75)

fig.show()
results=px.get_trendline_results(fig)
print(results)
results.px_fit_results.iloc[0].summary()

                                      px_fit_results
0  <statsmodels.regression.linear_model.Regressio...


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.046
Model:                            OLS   Adj. R-squared:                  0.045
Method:                 Least Squares   F-statistic:                     110.6
Date:                Sun, 11 Dec 2022   Prob (F-statistic):           2.69e-25
Time:                        00:07:11   Log-Likelihood:                -45797.
No. Observations:                2302   AIC:                         9.160e+04
Df Residuals:                    2300   BIC:                         9.161e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -6.247e+07   1.16e+07     -5.382      0.000   -8.52e+07   -3.97e+07
x1          1.098e+06   1.04e+05     10.517      0.000    8.93e+05     1.3e+06
==============================================================================
Omnibus:                     2666.817   Durbin-Watson:                   1.188
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           491078.277
Skew:                           5.708   Prob(JB):                         0.00
Kurtosis:                      73.637   Cond. No.                         586.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

###4. Domestic_Lifetime_Gross vs Rank
Next, we create a scatterplot, its histogram representation and regression line that models the relationship between a movie's domestic lifetime gross and its ranking.

In [ ]:
fig = px.scatter(BOXOFFICE_DF, x="Domestic_Lifetime_Gross", y="Rank", color="Rank", title="Fig4. Domestic Lifetime Gross vs Ranking", hover_name="Title", size_max=60, trendline="ols", marginal_x="histogram")
fig.show()

###5. Foreign Lifetime Gross vs Ranking
Finally, we create a scatterplot and regression line that captures the relationship between a movie's foreign lifetime gross and its ranking.

In [ ]:
fig = px.scatter(BOXOFFICE_DF, x="Foreign_Lifetime_Gross", y="Rank", color="Rank", title="Fig5. Foreign Lifetime Gross vs Ranking",  hover_name="Title", size_max=60, trendline="ols")
fig.update_yaxes(range=[0, 200])
fig.show()

# Conclusion + Key Insights


Circling back to the central question, we set out to investigate how factors such as rating, domestic lifetime gross, foreign lifetime gross and run-time affect a movie’s performance as measured by worldwide lifetime revenue (gross) and a movie's overall performance on the charts as measured by its ranking. Here are some key insights we obtained:


###Movie Ratings Distribution


From Figure 1, we see that the most common rating for the top 200 movies is a score of 8.1, while a score of 8.9 has the lowest movie count. Interestingly, no movie achieved a rating of 9.1

###Run-time vs Revenue
From Figure 2, we notice that in the range 60-240, the data is skewed to the left. Specifically, movie runtimes are heavily distributed in the range 95-105 minutes. In addition, based on figure 3, we can see that although there seems to be a positive relationship between runtime and movie revenue, the r squared coefficient tells a different story. An r squared value of 0.0046, as referenced by the table below, signals that there is no relationship between runtime and revenue

Absolute Value | Strenght of Relationship
-|-
r < 0.3|None or very weak
0.3 < r < 0.5| Weak
0.5 < r < 0.7| Moderate
r > 0.7 | Strong

###Domestic Lifetime Gross vs Ranking
From Figure 4, we see that domestic lifetime gross values are densely populated in the range 200-350million dollars. In addition, the regression line produces an r squared value of 0.46, which signals a weakly negative relationship between domestic lifetime gross and movie ranking.

###Foreign Lifetime Gross vs Ranking
From Figure 5, the regression line outputs an r squared value of 0.52, which signals a moderately negative relationship between foreign lifetime gross and movie ranking. This means that when foreign lifetime gross increases, a movie is expected to be rank higher on the charts. Since 0.52>0.46, foreign lifetime gross correlates slightly stronger with ranking than domestic gross, which indicates that foreign lifetime gross is a slightly better indicator of a movie's performance in the ranking charts.